File: 2025/python/two.py
ranges = list(
    map(
        lambda x: tuple(map(int, x.split("-"))),
        open("real.input", "r").read().split(","),
    )
)


def part_one(intervals: list[tuple]):
    invalid_id_sum = 0
    for interval in intervals:
        for number in range(interval[0], interval[1] + 1):
            number_string = str(number)
            number_length = len(number_string)
            if number_length % 2 != 0:
                continue
            if (
                number_string[: number_length // 2]
                == number_string[number_length // 2 :]
            ):
                invalid_id_sum += number
    print(f"solution1: {invalid_id_sum}")


def part_two(intervals: list[tuple]):
    invalid_id_sum = 0
    for interval in intervals:
        for number in range(interval[0], interval[1] + 1):
            number_string = str(number)
            number_length = len(number_string)
            for part_length in range(1, len(number_string) // 2 + 1):
                duplicates = set()
                for start_index in range(0, number_length, part_length):
                    devided_number = number_string[
                        start_index : start_index + part_length
                    ]
                    duplicates.add(devided_number)

                if len(duplicates) == 1:
                    invalid_id_sum += number
                    break

    print(f"solution2: {invalid_id_sum}")


part_one(ranges)
part_two(ranges)

File: 2025/rust/src/bin/two.rs
use std::ops::RangeInclusive;

fn main() {
    let ranges = include_str!("real.input")
        .split(",")
        .map(|x| (x.split("-").nth(0).unwrap(), x.split("-").nth(1).unwrap()))
        .map(|(x, y)| (x.trim(), y.trim()))
        .map(|(x, y)| (x.parse::<i64>().unwrap(), y.parse::<i64>().unwrap()))
        .map(|(x, y)| x..=y)
        .collect::<Vec<RangeInclusive<i64>>>();
    println!("{:?}", ranges);
    part_one(&ranges);
    part_two(&ranges);
}

fn part_one(ranges: &Vec<RangeInclusive<i64>>) {
    let mut invalid_id_sum = 0;
    let mut number_string: String;
    let mut number_length: usize;
    let mut first_half: String;
    let mut second_half: String;
    for range in ranges {
        for number in range.clone() {
            number_string = number.to_string();
            number_length = number_string.len();

            if number_length % 2 != 0 {
                continue;
            }

            first_half = number_string[..number_length.div_euclid(2)].to_string();
            second_half = number_string[number_length.div_euclid(2)..].to_string();

            if first_half == second_half {
                invalid_id_sum += number
            }
        }
    }
    println!("solution1={}", invalid_id_sum)
}

fn part_two(ranges: &Vec<RangeInclusive<i64>>) {
    let mut invalid_id_sum = 0;
    let mut number_string;
    let mut number_length;
    for range in ranges {
        for number in range.clone() {
            number_string = number.to_string();
            number_length = number_string.len();
            for part_length in 1..=number_length.div_euclid(2) {
                if number_length % part_length != 0 {
                    continue;
                }
                if number_string[..part_length].repeat(number_length / part_length) == number_string
                {
                    invalid_id_sum += number;
                    break;
                }
            }
        }
    }
    println!("solution2={}", invalid_id_sum)
}
